src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -2,18 +2,171 @@
+        while (entry != null) {
+            if (entry.getKey() == null || entry.getKey().equals(key)) {
+                return entry;
+            }
+            entry = entry.getNext();
+        }
+        return null;
+    }
+
+    private Node<K, V> findNullKeyNode(Node<K, V> entry) {
+        while (entry != null) {
+            if (entry.getKey() == null) {
+                return entry;
+            }
+            entry = entry.getNext();
+        }
+        return null;
+    }
+
+    private void addToBucket(Node<K, V> newEntry, int index) {
+        Node<K, V> entry = buckets[index];
+        if (entry != null) {
+            while (entry.getNext() != null) {
+                entry = entry.getNext();
+            }
+            entry.setNext(newEntry);
+        } else {
+            buckets[index] = newEntry;
+        }
+    }
+
+    private int getIndex(K key) {
+        if (key == null) {
+            return getNullKeyIndex();
+        }
+        int hashCode = key.hashCode();
+        int index = hashCode % buckets.length;
+        return index < 0 ? index + buckets.length : index;
+    }
+
+    private int getIndex(K key, int capacity) {
+        int hashCode = key.hashCode();
+        int index = hashCode % capacity;
+        return index < 0 ? index + capacity : index;
+    }
+
+    private int getNullKeyIndex() {
+        return buckets.length - 1;
+    }
+
+    private void resizeIfNeeded() {
+        if ((float) size / buckets.length >= DEFAULT_LOAD_FACTOR) {
+            int newCapacity = buckets.length * 2;
+            Node<K, V>[] newBuckets = new Node[newCapacity];
+            for (Node<K, V> node : buckets) {
+                while (node != null) {
+                    Node<K, V> next = node.getNext();
+                    int newIndex = getIndex(node.getKey(), newCapacity);
+                    node.setNext(newBuckets[newIndex]);
+                    newBuckets[newIndex] = node;
+                    node = next;
+                }
+            }
+            buckets = newBuckets;
+        }
+    }
+
+    private V getValueForNullKey() {
+        int index = getNullKeyIndex();
+        Node<K, V> entry = buckets[index];
+        Node<K, V> nullKeyNode = findNullKeyNode(entry);
+        return (nullKeyNode != null) ? nullKeyNode.getValue() : null;
+    }
+
+    private static class Node<K, V> {
+        private final K key;
+        private V value;
+        private Node<K, V> next;
+
+        public Node(K key, V value) {
+            this.key = key;
+            this.value = value;
+        }
+
+        public K getKey() {
+            return key;
+        }
+
+        public V getValue() {
+            return value;
+        }
+
+        public void setValue(V value) {
+            this.value = value;
+        }
+
+        public Node<K, V> getNext() {
+            return next;
+        }
+
+        public void setNext(Node<K, V> next) {
+            this.next = next;
+        }
```

Comment:
```suggestion
        private Node(K key, V value) {
            this.key = key;
            this.value = value;
        }
```
As described in [common mistakes](https://mate-academy.github.io/jv-program-common-mistakes/java-core/hash-map/hash-map#dont-use-getters-and-setters-in-class-node), let's not use getters and setters here